iceoryx_hoofs/utility/include/iox/detail/deprecation_marker.hpp

```diff
@@ -0,0 +1,131 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef IOX_HOOFS_UTILITY_DEPRECATION_MARKER_HPP
+#define IOX_HOOFS_UTILITY_DEPRECATION_MARKER_HPP
+
+#include "iceoryx_versions.hpp"
+
+namespace iox
+{
+namespace detail
+{
+struct DeprecationMarker
+{
+};
+} // namespace detail
+} // namespace iox
+
+// NOLINTJUSTIFICATION there is no other way to create the intended functionality for the deprecation marker
+// NOLINTBEGIN(cppcoreguidelines-macro-usage)
+
+#define IOX_INTERNAL_NEXT_DEPRECATED_VERSION 3
+
+#define IOX_INTERNAL_DEPRECATED_STINGIFY_HELPER_EXPAMD(NUM) #NUM
+#define IOX_INTERNAL_DEPRECATED_STINGIFY_HELPER(NUM) IOX_INTERNAL_DEPRECATED_STINGIFY_HELPER_EXPAMD(NUM)
+
+// clang-format off
+static_assert(ICEORYX_VERSION_MAJOR < IOX_INTERNAL_NEXT_DEPRECATED_VERSION,
+    "The iceoryx major version changed to v" IOX_INTERNAL_DEPRECATED_STINGIFY_HELPER(ICEORYX_VERSION_MAJOR) "!\n"
+    "The following steps need to be done to fix this error: \n"
+    " - increment 'IOX_INTERNAL_NEXT_DEPRECATED_VERSION'\n"
+    " - update 'IOX_INTERNAL_DEPRECATED_SINCE_V" IOX_INTERNAL_DEPRECATED_STINGIFY_HELPER(ICEORYX_VERSION_MAJOR)
+        " to call 'IOX_INTERNAL_DEPRECATED_SINCE(VERSION, MESSAGE)'\n"
+    " - update 'IOX_INTERNAL_DEPRECATED_HEADER_SINCE_V" IOX_INTERNAL_DEPRECATED_STINGIFY_HELPER(ICEORYX_VERSION_MAJOR)
+        " to call 'IOX_INTERNAL_DEPRECATED_HEADER_SINCE(VERSION, MESSAGE)'");
+
+// clang-format on
+
+
+// BEGIN IOX_DEPRECATED_HEADER_SINCE macros
+
+// when the namespace is called 'header' the warning will be "warning: 'header' is deprecated: Deprecated since v#.0
+// ..." with the file and line where the 'IOX_DEPRECATED_HEADER_SINCE' macro is used
+#define IOX_INTERNAL_DEPRECATED_HEADER_SINCE(VERSION, MESSAGE)                                                         \
+    namespace iox                                                                                                      \
+    {                                                                                                                  \
+    namespace detail                                                                                                   \
+    {                                                                                                                  \
+    namespace                                                                                                          \
+        [[deprecated("Deprecated since v" #VERSION ".0 and will be removed at a later version! " MESSAGE)]] header     \
```

Comment:
Ah, now I understand it :) Ignore my comment before.

Reply:
Have you tried it out and could add the compiler output? I'm not sure if this will lead to showing the user the wrong namespace `detail`. What we want to communicate is the deprecation of the **header**.

Reply:
Here the output when the deprecation version is set to 2 in `iceoryx_hoofs_types.hpp` and `iox::cxx:byte_t` is used in `test_utility_deprecation_marker.cpp`

Warning for the header deprecation
```console
In file included from /home/.../iceoryx/iceoryx_hoofs/test/moduletests/test_utility_deprecation_marker.cpp:22:
/home/.../iceoryx/iceoryx_hoofs/legacy/include/iceoryx_hoofs/iceoryx_hoofs_types.hpp:23:1: warning: 'header' is deprecated: Deprecated since v2.0 and will be removed at a later version! Please include 'iox/iceoryx_hoofs_types.hpp' instead. [-Wdeprecated-declarations]
```

Warning for the `byte_t` deprecation
```console
/home/.../iceoryx/iceoryx_hoofs/test/moduletests/test_utility_deprecation_marker.cpp:26:11: warning: 'byte_t' is deprecated: Deprecated since v2.0 and will be removed at a later version! Please use 'iox::byte' instead. [-Wdeprecated-declarations]
```